windows/src/test/unit-tests/kmconvert/Keyman.Developer.System.Test.KMConvertParametersTest.pas

```diff
@@ -0,0 +1,184 @@
+  Assert.IsFalse(k.CheckParams(['template', '-id', 'FooBar']));
+  Assert.IsFalse(k.CheckParams(['template', '-id', 'foo$']));
+  Assert.IsFalse(k.CheckParams(['template', '-id', 'foo-bar']));
+  Assert.IsFalse(k.CheckParams(['template', '-id', '0123']));
+  Assert.IsFalse(k.CheckParams(['template', '-id', '0foo']));
+  Assert.IsTrue(k.CheckParams(['template', '-id', 'foo_123']));
+end;
+
+procedure TKMConvertParametersTest.TestName;
+begin
+  Assert.Pass;  // no validation
+end;
+
+procedure TKMConvertParametersTest.TestCopyright;
+begin
+  Assert.Pass;  // no validation
+end;
+
+procedure TKMConvertParametersTest.TestFullCopyright;
+begin
+  Assert.Pass;  // no validation
+end;
+
+procedure TKMConvertParametersTest.TestVersion;
+var
+  k: TKMConvertParameters;
+begin
+  k.OnOutputText := procedure(s: string) begin Log(TLogLevel.Information, s); end;
+  Assert.IsTrue(k.CheckParams(['template', '-version', '1.2.3']));
+  Assert.IsFalse(k.CheckParams(['template', '-version', '1.2.3a']));
+  Assert.IsFalse(k.CheckParams(['template', '-version', '1.0-alpha']));
+  Assert.IsFalse(k.CheckParams(['template', '-version', 'evergreen']));
+end;
+
+procedure TKMConvertParametersTest.TestBCP47Tags;
+var
+  k: TKMConvertParameters;
+  Found: Boolean;
+begin
+  k.OnOutputText := procedure(s: string) begin Log(TLogLevel.Information, s); end;
+  Assert.IsTrue(k.CheckParams(['template', '-languages', 'en fr']));
+  Assert.IsFalse(k.CheckParams(['template', '-languages', 'english']));
+  Assert.IsFalse(k.CheckParams(['template', '-languages', '$$$']));
+
+  Found := False;
+  k.OnOutputText :=
+    procedure(s: string) begin
+      Log(TLogLevel.Information, s);
+      Assert.AreEqual(s, 'Warning: -languages: ''en-us'' is a valid tag but is not canonical: it should be ''en''');
+      Found := True;
+    end;
+  Assert.IsTrue(k.CheckParams(['template', '-languages', 'en-us']));
+  Assert.IsTrue(Found);
+end;
+
+procedure TKMConvertParametersTest.TestAuthor;
+begin
+  Assert.Pass;  // no validation
+end;
+
+procedure TKMConvertParametersTest.TestTargets;
+var
+  k: TKMConvertParameters;
+begin
+  k.OnOutputText := procedure(s: string) begin Log(TLogLevel.Information, s); end;
+  Assert.IsTrue(k.CheckParams(['template', '-targets', 'any']));
+  Assert.IsTrue(k.CheckParams(['template', '-targets', 'windows macos']));
+  Assert.IsFalse(k.CheckParams(['template', '-targets', 'pdp-11']));
+end;
+
+procedure TKMConvertParametersTest.TestModelIdAuthor;
+var
+  k: TKMConvertParameters;
+begin
+  k.OnOutputText := procedure(s: string) begin Log(TLogLevel.Information, s); end;
+  Assert.IsTrue(k.CheckParams(['template', '-id-author', 'myself']));
+  Assert.IsTrue(k.CheckParams(['template', '-id-author', 'my_self']));
+  Assert.IsFalse(k.CheckParams(['template', '-id-author', 'my-self']));
+  Assert.IsFalse(k.CheckParams(['template', '-id-author', 'Myself']));
+  Assert.IsFalse(k.CheckParams(['template', '-id-author', '$myself$']));
```

Comment:
Are there restrictions on valid `-id-author values`?

https://help.keyman.com/developer/current-version/context/kmconvert#toc-lexical-model-parameters

Reply:
Yes, the -id-author and -id-uniq values must be valid tokens, i.e. lower case a-z, 0-9, and underscore. Should start with alpha or underscore. The `author.bcp47.uniq` value has the following regex: 

```
^[a-z_][a-z0-9_]*\.[a-z_][a-z0-9_-]*\.[a-z_][a-z0-9_]*$
```

Reply:
How about a test?
```
Assert.IsTrue(k.checkParams(['template', 'id-author', '_myself']));
```

Reply:
Ah, missed this comment before I merged! Next time I guess?